File: apps/desktop/test/server-url.test.mjs
import assert from "node:assert/strict";
import { test } from "node:test";
import { sameOrigin, serverCandidates } from "../src/server-url.mjs";

test("keeps a full address and drops its path", () => {
  assert.deepEqual(serverCandidates(" https://kru.example.com/login?next=/app "), ["https://kru.example.com"]);
  assert.deepEqual(serverCandidates("http://192.168.1.20:3000/"), ["http://192.168.1.20:3000"]);
});

test("tries https first for public names", () => {
  assert.deepEqual(serverCandidates("kru.example.com"), ["https://kru.example.com", "http://kru.example.com"]);
});

test("tries http first for this machine and private networks", () => {
  assert.deepEqual(serverCandidates("localhost:3000"), ["http://localhost:3000", "https://localhost:3000"]);
  assert.deepEqual(serverCandidates("10.0.0.5:3000"), ["http://10.0.0.5:3000", "https://10.0.0.5:3000"]);
  assert.deepEqual(serverCandidates("nas.local"), ["http://nas.local", "https://nas.local"]);
});

test("refuses what isn't a web address", () => {
  assert.throws(() => serverCandidates(""), /Enter the address/);
  assert.throws(() => serverCandidates("file:///etc/passwd"), /http:\/\/ or https:\/\//);
  assert.throws(() => serverCandidates("https://me:pw@kru.example.com"), /username and password/);
  assert.throws(() => serverCandidates("https://"), /doesn't look like an address/);
});

test("matches origins exactly", () => {
  assert.equal(sameOrigin("https://kru.example.com/app/t/1", "https://kru.example.com"), true);
  assert.equal(sameOrigin("https://kru.example.com.evil.test/", "https://kru.example.com"), false);
  assert.equal(sameOrigin("http://kru.example.com/", "https://kru.example.com"), false);
  assert.equal(sameOrigin("not a url", "https://kru.example.com"), false);
});

File: apps/android/app/src/test/kotlin/app/krubot/android/ServerUrlTest.kt
package app.krubot.android

import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertTrue
import org.junit.Test

/** Mirrors apps/desktop/test/server-url.test.mjs, so both apps read an address the same way. */
class ServerUrlTest {
    @Test
    fun `an address with a scheme is tried as given, without its path`() {
        assertEquals(listOf("https://kru.example.com"), ServerUrl.candidates("https://kru.example.com/app/t/123"))
        assertEquals(listOf("http://192.168.1.20:3000"), ServerUrl.candidates("http://192.168.1.20:3000/"))
    }

    @Test
    fun `a public host without a scheme tries https first`() {
        assertEquals(listOf("https://kru.example.com", "http://kru.example.com"), ServerUrl.candidates("kru.example.com"))
    }

    @Test
    fun `this machine and private networks try http first`() {
        assertEquals(listOf("http://localhost:3000", "https://localhost:3000"), ServerUrl.candidates("localhost:3000"))
        assertEquals(listOf("http://192.168.1.20:3000", "https://192.168.1.20:3000"), ServerUrl.candidates("192.168.1.20:3000"))
        assertEquals(listOf("http://kru.local", "https://kru.local"), ServerUrl.candidates("kru.local"))
        assertEquals(listOf("http://nas", "https://nas"), ServerUrl.candidates("nas"))
    }

    @Test
    fun `default ports are dropped and others kept`() {
        assertEquals(listOf("http://kru.example.com"), ServerUrl.candidates("http://kru.example.com:80"))
        assertEquals(listOf("https://kru.example.com:8443"), ServerUrl.candidates("https://kru.example.com:8443"))
    }

    @Test
    fun `bad input names its reason`() {
        assertEquals(ServerUrl.Reason.EMPTY, reason("  "))
        assertEquals(ServerUrl.Reason.SCHEME, reason("ftp://kru.example.com"))
        assertEquals(ServerUrl.Reason.CREDENTIALS, reason("https://me:secret@kru.example.com"))
        assertEquals(ServerUrl.Reason.INVALID, reason("http://"))
    }

    @Test
    fun `sameOrigin compares scheme, host and port`() {
        assertTrue(ServerUrl.sameOrigin("https://kru.example.com/app", "https://kru.example.com"))
        assertTrue(ServerUrl.sameOrigin("https://KRU.example.com:443/x", "https://kru.example.com"))
        assertFalse(ServerUrl.sameOrigin("http://kru.example.com/app", "https://kru.example.com"))
        assertFalse(ServerUrl.sameOrigin("https://kru.example.com:8443/", "https://kru.example.com"))
        assertFalse(ServerUrl.sameOrigin("not a url", "https://kru.example.com"))
        assertFalse(ServerUrl.sameOrigin(null, "https://kru.example.com"))
    }

    private fun reason(input: String): ServerUrl.Reason = try {
        ServerUrl.candidates(input)
        error("expected $input to be rejected")
    } catch (failure: ServerUrl.Invalid) {
        failure.reason
    }
}
